python-package/lightgbm/basic.py

```diff
@@ -591,6 +594,75 @@ def _load_pandas_categorical(file_name=None, model_str=None):
         return None
 
 
+class Sequence:
+    """
+    Generic data access interface.
+
+    Object should support the following operations:
+
+        # Get total row number.
+        >>> len(seq)
+        # Random access by row index. Use for data sampling.
+        >>> seq[10]
+        # Range data access. Use to read data in batch when constructing Dataset.
+        >>> seq[0:100]
+        # Optionally specify batch_size to control range data read size.
+        >>> seq.batch_size
+
+    With random access, data sampling does not need to go through all data.
+    With range data access, there's no need to read all data into memory thus
+    reduce memory usage.
+    """
+
+    batch_size = 4096  # Defaults to read 4K rows in each batch.
+
+    @staticmethod
+    def is_class(obj) -> bool:
+        """Check if objection is instance of class Sequence.
+
+        Args:
+        -------
+            obj ([any]): object to be checked
+
+        Returns
+        -------
+            [bool]: is Sequence class
+        """
+        if isinstance(obj, list) or hasattr(obj, "getformat"):
+            return False
+        return hasattr(obj, "__getitem__") and hasattr(obj, "__len__")
+
+    def __getitem__(self, idx: Union[int, slice]) -> np.ndarray:
+        """Return data for given row index.
+
+        A basic implementation should look like this:
+
+        .. code-block:: python
+
+            if isinstance(idx, numbers.Integral):
+                return self.__get_one_line__(idx)
+            elif isinstance(idx, slice):
+                return np.stack(self.__get_one_line__(i) for i in range(idx.start, idx.stop))
+            else:
+                raise TypeError("Sequence index must be integer or slice, got {}".format(type(idx)))
+
+        Parameters
+        ----------
+        idx : int, slice[int]
+            Item index.
+
+        Returns
+        -------
+        result : numpy 1-D array, numpy 2-D array
+            1-D array if idx is int, 2-D array if idx is slice
+        """
+        raise NotImplementedError("remove this line if subclassing")
+
+    def __len__(self) -> int:
+        """Return row count of this sequence."""
+        raise NotImplementedError
```

Comment:
```suggestion
        raise NotImplementedError("Sub-classes of lightgbm.Sequence must implement __len__()")
```

Could this have a more specific error message, similar to the one I just suggested for `__getitem__()`? I think that will help debugging cases where someone uses code that subclassed `Sequence` but failed to implement this.